File: 1011-Ship-With-In-Days/1101.py
class Solution:
  def shipWithInDays(self, weights: List[int], D: int) -> int:
    left, right = max(weights), sum(weights)
    while left < right:
      mid = (left + right) // 2
      need, cur = 1, 0
      for weight in weights:
        if cur + weight > mid:
          need += 1
          cur = 0
        cur += weight

      if need <= D:
        right = mid
      else:
        left = mid + 1

    return left

File: 1001-1100/1011-Ship-With-In-Days/1101.java
class Solution {
  public int shipWithInDays(int[] weights, int D) {
    int left = Arrays.stream(weights).max().getAsInt(), right = Arrays.stream(weights).sum();
    while (left < right) {
      int mid = (left + right) / 2;
      int need = 1, cur = 0;
      for (int weight : weights) {
        if (cur + weight > mid) {
          ++need;
          cur = 0;
        }
        cur += weight;
      }
      if (need <= D) {
        right = mid;
      } else {
        left = mid + 1;
      }
    }
    return left;
  }
}